src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java

```diff
@@ -598,13 +579,92 @@ private void generateQuotedMethod(ClassBuilder clb) {
         clb.withMethod(NAME_METHOD_QUOTED, CodeReflectionSupport.MTD_Quoted, ACC_PUBLIC + ACC_FINAL, new MethodBody(new Consumer<CodeBuilder>() {
             @Override
             public void accept(CodeBuilder cob) {
+                // bytecode for double-checked locking (copied from the compiler)
+                Label l1 = cob.newLabel();
+                Label l2 = cob.newLabel();
+                Label target = cob.newLabel();
+                Label from1 = cob.newLabel();
+                Label to1 = cob.newLabel();
+                Label from2 = cob.newLabel();
+                Label to2 = cob.newLabel();
                 cob.aload(0)
-                   .getfield(lambdaClassEntry.asSymbol(), quotedInstanceFieldName, CodeReflectionSupport.CD_Quoted)
-                   .areturn();
+                        .getfield(lambdaClassEntry.asSymbol(), quotedInstanceFieldName, CodeReflectionSupport.CD_Quoted)
+                        .astore(1)
+                        .aload(1)
+                        .ifnonnull(l1)
+                        .aload(0)
+                        .dup()
+                        .astore(2)
+                        .monitorenter()
+                        .labelBinding(from1)
+                        .aload(0)
+                        .getfield(lambdaClassEntry.asSymbol(), quotedInstanceFieldName, CodeReflectionSupport.CD_Quoted)
+                        .astore(1)
+                        .aload(1)
+                        .ifnonnull(l2);
+                createQuotedObj(cob);
```

Comment:
I find this method confusing because we are performing DCL over the quoted instance and not the code model. I can see why you did that.

Can we refine this in a subsequent PR to do so on the model and then we can create `Quoted` instance each time? Later we can turn `Quoted` into a value class, and for now we can mark it as value-based. Then i think that allows us to adjust later if we so wish to store the code model in a weak reference